### DISCUSSION TITLE
Query time with populate & responseCache is a bit weird

### DISCUSSION THREAD

**Discussion by mvuckovic** [mikro-orm/mikro-orm#5816] [Q&A]
Hello! 

I have some public API endpoints so I was looking into leveraging mikro-orm resultCache in order to get better RPS on the API but I'm super confused on numbers in different combinations:

## Setup (I can provide full entity definitions if required)
- Next.js with route handlers
- Mikro-orm 6.2.6
- bench-rest for load testing
- loadStrategy: "joined"

I basically have 4 entities (assortment, assortment-translation, product, product-translation).
There's also an API flag includeProducts which adds additional populates.

Given following query:
```
const data = await em.find(
  Assortment,
  {},
  {
    populate: includeProducts
      ? ["translations", "products", "products.translations"]
      : ["translations"],
    filters: { language: { language } },
  },
);
```

1. Getting just assortments + translations, no cache, avg ~150 RPS
2. Getting just assortments + translations with cache: 60000, avg ~250 RPS
3. Getting assortments + translations + products + product translations, no cache, avg ~120 RPS
4. Getting assortments + translations + products + product translations with cache, avg ~30 RPS

Now if I populate products + product translations separately (it seems there's no cache option on populate call)

```
const data = await em.find(
  Assortment,
  {},
  {
    populate: ["translations"],
    filters: { language: { language } },
  },
);

if (includeProducts) {
  await em.populate(data, ["products", "products.translations"], {
    filters: { language: { language } },
  });
}
```
1. Getting assortments + translations, and then products + product translations, no cache, avg ~120 RPS
2. Getting assortments + translations cached, and then products + product translations, no cache, avg ~150 RPS

And finally if I fetched them both separately and cache them separately

Note: I wasn't able to find a way on how to populate assortment products after loading them separately.

```
const data = await em.find(
  Assortment,
  {},
  {
    populate: ["translations"],
    filters: { language: { language } },
    //cache: 60000,
  },
);

let products = [];

if (includeProducts) {
  products = await em.find(
    Product,

    {
      assortment: data,
    },
    {
      populate: ["translations"],
      filters: { language: { language } },
    },
  );
}
```
1. Getting assortments + translations no cache, and then products + product translations, no cache, avg ~120 RPS
2. Getting assortments + translations cached, and then products + product translations, cached, avg ~110 RPS

I guess my question boils down to: Are these values expected? Especially the first one which seems slow with populate + cache.

**Comment by B4nan** [comment]
I can't help with perf issues without seeing them on my end, that would be a complete guessing game.

Based on the property names, it feels like you are trying to join two to-many relations (`products.translations` sounds like you are even doing that on two levels), so maybe try select-in strategy instead of the joined one, since this can explode and produce a huge cartesian product.

**Reply by mvuckovic** [reply]
I understand, first I wanted to find out is such results are expected or not. 
I can create a code sample git repository containing everything to run it.

I did try with `select-in` as well but the results seem very similar,

No-cache version yields around ~115 RPS while the one with responseCache yields around ~30 RPS
```
[query] select "a0".* from "assortment" as "a0" [took 1 ms, 1 result]
[query] select "a0".* from "assortment_translation" as "a0" where "a0"."language_id" in ('de', 'en') and "a0"."assortment_id" in ('117') order by "a0"."priority" asc [took 1 ms, 2 results]
[query] select "p0".* from "product" as "p0" where "p0"."assortment_id" in ('117') [took 1 ms, 6 results]
[query] select "p0".* from "product_translation" as "p0" where "p0"."language_id" in ('de', 'en') and "p0"."product_id" in ('3055', '3057', '3059', '2816', '3056', '3058') order by "p0"."priority" asc [took 1 ms, 12 results]
```

The joined one results in 24 results as the data sample is small
```
[query] select "a0".*, "t1"."id" as "t1__id", "t1"."created_at" as "t1__created_at", "t1"."updated_at" as "t1__updated_at", "t1"."assortment_id" as "t1__assortment_id", "t1"."language_id" as "t1__language_id", "t1"."name" as "t1__name", "t1"."priority" as "t1__priority", "t1"."description" as "t1__description", "p2"."id" as "p2__id", "p2"."created_at" as "p2__created_at", "p2"."updated_at" as "p2__updated_at", "p2"."assortment_id" as "p2__assortment_id", "p2"."languages" as "p2__languages", "p2"."attributes" as "p2__attributes", "t3"."id" as "t3__id", "t3"."created_at" as "t3__created_at", "t3"."updated_at" as "t3__updated_at", "t3"."product_id" as "t3__product_id", "t3"."language_id" as "t3__language_id", "t3"."name" as "t3__name", "t3"."description" as "t3__description", "t3"."priority" as "t3__priority" from "assortment" as "a0" left join "assortment_translation" as "t1" on "a0"."id" = "t1"."assortment_id" and "t1"."language_id" in ('de', 'en') left join "product" as "p2" on "a0"."id" = "p2"."assortment_id" left join "product_translation" as "t3" on "p2"."id" = "t3"."product_id" and "t3"."language_id" in ('de', 'en') order by "t1"."priority" asc, "t3"."priority" asc [took 6 ms, 24 results]
```

So from this testing it looks (at least to me) that resultCache works great with a single populate (where assortment + translations went from 150 -> 250 RPS with cache) but having more and nested populates results in slower response (assortment + translation + product + product translations went from 120 -> 30 RPS with cache)

Kind regards and thank you for such a quick response :)
